### DISCUSSION TITLE
Bluetooth dongle (TP-link UB500 / RTL8761B) support needed

### DISCUSSION THREAD

**Discussion by rahul-kurup** [Joshua-Riek/ubuntu-rockchip#94] [General]
I have a [TP-Link UB500 Nano](https://www.tp-link.com/in/home-networking/adapter/ub500/) dongle, which gets detected properly in the OS

<img width="583" alt="image" src="https://user-images.githubusercontent.com/29568606/228947694-757c94c8-522d-46e0-9177-3e46309fc6e6.png">

But the problem is, the scanning doesn't work and it always stays in this same state

<img width="556" alt="image" src="https://user-images.githubusercontent.com/29568606/228947517-983963ff-fb5a-4f9b-82cd-b7d2b83aa9cb.png">

`lsusb; dmesg | egrep -i 'blue|firm'` outputs:

```
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
[    5.358465] psci: PSCIv1.1 detected in firmware.
[    6.867404] arm-scmi firmware:scmi: SCMI Notifications - Core Enabled.
[    6.867434] arm-scmi firmware:scmi: SCMI Protocol v2.0 'rockchip:' Firmware version 0x0
[    6.868545] Bluetooth: Core ver 2.22
[    6.868559] Bluetooth: HCI device and connection manager initialized
[    6.868564] Bluetooth: HCI socket layer initialized
[    6.868567] Bluetooth: L2CAP socket layer initialized
[    6.868573] Bluetooth: SCO socket layer initialized
[    9.144206] arm-scmi firmware:scmi: Failed. SCMI protocol 17 not active.
[    9.150274] optee: probe of firmware:optee failed with error -22
[    9.225463] [BT_RFKILL]: bluetooth_platdata_parse_dt: get property: uart_rts_gpios = 122.
[    9.225480] [BT_RFKILL]: bluetooth_platdata_parse_dt: get property: BT,reset_gpio = 102.
[    9.225490] [BT_RFKILL]: bluetooth_platdata_parse_dt: get property: BT,wake_gpio = 22.
[    9.276244] mali fb000000.gpu: Capping CSF_FIRMWARE_TIMEOUT to CSF_FIRMWARE_PING_TIMEOUT
[    9.352726] platform regulatory.0: Direct firmware load for regulatory.db failed with error -2
[    9.360978] Bluetooth: hci0: CSR: Unbranded CSR clone detected; adding workarounds...
[   12.044089] Bluetooth: hci0: command 0x0c01 tx timeout
[   14.180378] Bluetooth: hci0: command 0x080f tx timeout
[   16.296696] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[   16.296710] Bluetooth: BNEP filters: protocol multicast
[   16.296730] Bluetooth: BNEP socket layer initialized
[   16.310324] Bluetooth: hci0: command 0x0c1b tx timeout
[   18.443769] Bluetooth: hci0: command 0x0c46 tx timeout
[   20.577122] Bluetooth: hci0: command 0x0c63 tx timeout
[   22.710442] Bluetooth: hci0: command 0x100b tx timeout
[   24.750610] mali fb000000.gpu: Loading Mali firmware 0x1010000
[   24.751901] mali fb000000.gpu: Mali firmware git_sha: 9b8db9aa05a7b4f3066fd75dbdff695262f891fe 
[   24.843739] Bluetooth: hci0: command 0x0c77 tx timeout
[   26.976979] Bluetooth: hci0: command 0x2024 tx timeout
[   29.110323] Bluetooth: hci0: command tx timeout
[  476.559484] Bluetooth: RFCOMM TTY layer initialized
[  476.559510] Bluetooth: RFCOMM socket layer initialized
[  476.559533] Bluetooth: RFCOMM ver 1.11
```

**Anyone know how to fix this?**

PS: I have tried 
- https://askubuntu.com/a/1371453
- https://www.youtube.com/watch?v=x69ADr-E2Lw

but still stuck in the same scan state where it fetches nothing!

**Comment by ironiris** [comment]
In my case, I've been using the mouse from the link I sent you.

**Reply by rahul-kurup** [reply]
@ironiris Looks like I cannot get anything delivered to India from AliExpress.
 Thanks for sharing the link :)

**Reply by rahul-kurup** [reply]
unfortunately AliExpress is not delivering to India right now

**Comment by rahul-kurup** [comment]
Hey @Joshua-Riek  I got a solution which _may_ work https://askubuntu.com/a/1371453

<img width="747" alt="image" src="https://user-images.githubusercontent.com/29568606/230177886-27c1d856-6696-47b9-9fbf-51d4f81d2d0a.png">

But I need to configure linux headers for op5 chipset `linux-headers-5.10.110-rockchip-rk3588` to satisfy the above code block

Do you have any idea on how to install the headers? 

PS: I've also found the source for linux headers here https://mirrors.shanghaitech.edu.cn/armbian/pool/main/l/linux-5.10.110-rockchip-rk3588/

**Reply by Joshua-Riek** [reply]
The headers should be included, but for some reason, it does not look like it was included. I'm really busy right now with work and should be able to send you the proper headers in a few hours.

**Reply by rahul-kurup** [reply]
@Joshua-Riek I found this thread https://bbs.archlinux.org/viewtopic.php?pid=2004851#p2004851 where support for UB500 is added, thus fixing the scan issue. Would it be possible for you to port this in your image?

**Reply by Joshua-Riek** [reply]
I don't think I can include that in my image, it's for the kernel version 5.15.x, and this board is using a poorly ported version of 5.10.110. 

I have the kernel headers available here: https://launchpadlibrarian.net/660302912/linux-headers-5.10.110-rockchip-rk3588_5.10.110~git20230406.88961a7-1_arm64.deb

Quick note, the link might break at some point, I'm in the process of packaging the kernel so everyone can receive updates with a simple apt upgrade command.

**Reply by rahul-kurup** [reply]
Understandable. I'll mess around more to see what can be done.
